### PR TITLE
Python CAN support

### PR DESCRIPTION
**Pull request by sorki**
###### Motivation for this change

CAN/CANOpen python stack

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by grahamc** [comment]
@GrahamcOfBorg eval

(a commit broke master, and it is now fixed.)

**Review by dotlambda — COMMENTED** [review]
Please put the derivations into spererate files inside `pkgs/development/python-modules` and use `callPackage`.

File: pkgs/top-level/python-packages.nix [hunk 1]
```diff
@@ -2012,6 +2012,83 @@ in {
     doCheck = false;
   });
 
+  can = buildPythonPackage rec {
+    rev = "2.0.0";
+    name = "can-${rev}";
```
Comment (dotlambda) [inline pkgs/top-level/python-packages.nix]: No `name`. Specify `pname` and `version` please.

File: pkgs/top-level/python-packages.nix [hunk 2]
```diff
@@ -2012,6 +2012,83 @@ in {
...
+      homepage = https://github.com/hardbyte/python-can;
+      description = "CAN support for Python";
+      license = licenses.lgpl3;
+      maintainers = with maintainers; [ sorki ];
+    };
+  };
+
+  canopen = buildPythonPackage rec {
+    rev = "v0.5.1";
+    name = "canopen-${rev}";
```
Comment (dotlambda) [inline pkgs/top-level/python-packages.nix]: See above.

File: pkgs/top-level/python-packages.nix [hunk 3]
```diff
@@ -2012,6 +2012,83 @@ in {
...
+      homepage = https://github.com/christiansandberg/canopen/;
+      description = "CANopen stack implementation";
+      license = licenses.lgpl3;
+      maintainers = with maintainers; [ sorki ];
+    };
+  };
+
+  canmatrix = buildPythonPackage rec {
+    rev = "0.6";
+    name = "canmatrix-${rev}";
```
Comment (dotlambda) [inline pkgs/top-level/python-packages.nix]: See above.

File: pkgs/top-level/python-packages.nix [hunk 4]
```diff
@@ -2012,6 +2012,83 @@ in {
     doCheck = false;
   });
 
+  can = buildPythonPackage rec {
+    rev = "2.0.0";
+    name = "can-${rev}";
+
+    src = pkgs.fetchFromGitHub {
```
Comment (dotlambda) [inline pkgs/top-level/python-packages.nix]: Please use `fetchPypi` if possible. This makes it easier to override the `src`.

File: pkgs/top-level/python-packages.nix [hunk 5]
```diff
@@ -2012,6 +2012,83 @@ in {
     doCheck = false;
   });
 
+  can = buildPythonPackage rec {
```
Comment (FRidh) [inline pkgs/top-level/python-packages.nix]: We are moving Python expressions out of `pkgs/top-level/python-packages.nix` into `pkgs/development/python-modules/<module>/default.nix`. 

Please move the expression there, and call it from `pkgs/top-level/python-packages.nix` using `callPackage ../development/python-modules/<package> { };`.

**Review by FRidh — CHANGES_REQUESTED** [review]

**Comment by sorki** [comment]
Should be ok now.

File: pkgs/development/python-modules/canopen/default.nix [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+    inherit pname version;
+    sha256 = "0806cykarpjb9ili3mf82hsd9gdydbks8532nxgz93qzg4zdbv2g";
+  };
+
+  # due to missing test/sample.eds
+  # https://github.com/christiansandberg/canopen/pull/57
+  doCheck = false;
+
+  propagatedBuildInputs =
+    [ nose
```
Comment (dotlambda) [inline pkgs/development/python-modules/canopen/default.nix]: `checkInputs`

File: pkgs/development/python-modules/can/default.nix [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+  name = "${pname}-${version}";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "1c6zfd29ck9ffdklfb5xgxvfl52xdaqd89isykkypm1ll97yk2fs";
+  };
+
+  propagatedBuildInputs =
+    [ nose
+      mock
```
Comment (dotlambda) [inline pkgs/development/python-modules/can/default.nix]: `checkInputs`

File: pkgs/development/python-modules/can/default.nix [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+, buildPythonPackage
+, fetchPypi
+, pyserial
+, nose
+, mock }:
+
+buildPythonPackage rec {
+  pname = "python-can";
+  version = "2.0.0";
+  name = "${pname}-${version}";
```
Comment (dotlambda) [inline pkgs/development/python-modules/can/default.nix]: no `name`

File: pkgs/development/python-modules/canmatrix/default.nix [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+, xlwt
+, xlrd
+, XlsxWriter
+, pyyaml
+, future }:
+
+buildPythonPackage rec {
+  pname = "canmatrix";
+  version = "0.6";
+  name = "${pname}-${version}";
```
Comment (dotlambda) [inline pkgs/development/python-modules/canmatrix/default.nix]: no `name`

File: pkgs/development/python-modules/canopen/default.nix [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+, buildPythonPackage
+, fetchPypi
+, nose
+, can
+, canmatrix }:
+
+buildPythonPackage rec {
+  pname = "canopen";
+  version = "0.5.1";
+  name = "${pname}-${version}";
```
Comment (dotlambda) [inline pkgs/development/python-modules/canopen/default.nix]: no `name`

File: pkgs/development/python-modules/canopen/default.nix [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "0806cykarpjb9ili3mf82hsd9gdydbks8532nxgz93qzg4zdbv2g";
+  };
+
+  # due to missing test/sample.eds
+  # https://github.com/christiansandberg/canopen/pull/57
+  doCheck = false;
```
Comment (dotlambda) [inline pkgs/development/python-modules/canopen/default.nix]: Thanks a lot for opening the PR! As long as no new version is published on PyPI, you could also rely on `fetchFromGitHub`.

**Comment by dotlambda** [comment]
@GrahamcOfBorg build python2.pkgs.canopen python3.pkgs.canopen

**Comment by dotlambda** [comment]
The following enabled `canmatrix` tests. Because PyPI doesn't ship the `test/` directory, `fetchFromGitHub` is used.
```diff
diff --git a/pkgs/development/python-modules/canmatrix/default.nix b/pkgs/development/python-modules/canmatrix/default.nix
index c62c43dc2df..6409010aa3e 100644
--- a/pkgs/development/python-modules/canmatrix/default.nix
+++ b/pkgs/development/python-modules/canmatrix/default.nix
@@ -1,7 +1,7 @@
 { lib
 , stdenv
 , buildPythonPackage
-, fetchPypi
+, fetchFromGitHub
 , lxml
 , xlwt
 , xlrd
@@ -14,9 +14,11 @@ buildPythonPackage rec {
   version = "0.6";
   name = "${pname}-${version}";
 
-  src = fetchPypi {
-    inherit pname version;
-    sha256 = "1jv0ry3qrrkjc32a0lwznf7ix8wl2s5lccqsqcyxnr89qbkjapqb";
+  src = fetchFromGitHub {
+    owner = "ebroecker";
+    repo = pname;
+    rev = version;
+    sha256 = "1lb0krhchja2jqfsh5lsfgmqcchs1pd38akvc407jfmll96f4yqz";
   };
 
   propagatedBuildInputs =
@@ -28,6 +30,11 @@ buildPythonPackage rec {
       future
     ];
 
+  checkPhase = ''
+    cd test
+    python ./test.py
+  '';
+
   meta = with lib; {
     homepage = https://github.com/ebroecker/canmatrix;
     description = "Support and convert several CAN (Controller Area Network) database formats .arxml .dbc .dbf .kcd .sym fibex xls(x)";
```
However, I don't know what the expected outcome is. Their travis build more or less show the same output as this does.

**Comment by sorki** [comment]
Had to disable doCheck in `canopen` now, rest looks fine, I would let `canmatrix` run its test even if they only ouput diffs if they ever decide to return exit code according to that.

File: pkgs/development/python-modules/canmatrix/default.nix [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+  src = fetchFromGitHub {
+    owner = "ebroecker";
+    repo = pname;
+    rev = version;
+    sha256 = "1lb0krhchja2jqfsh5lsfgmqcchs1pd38akvc407jfmll96f4yqz";
+  };
+
+  checkPhase = ''
+    cd test
+    python ./test.py
```
Comment (FRidh) [inline pkgs/development/python-modules/canmatrix/default.nix]: `${python.interpreter`. This is needed e.g. for `pypy`

**Comment by FRidh** [comment]
@GrahamcOfBorg build python2.pkgs.can python2.pkgs.canmatrix python3.pkgs.can python3.pkgs.canmatrix

**Review by FRidh — CHANGES_REQUESTED** [review]

**Pull request merged**